### PR TITLE
expose site_status in UserProfileSerializer and /site/

### LINKED ISSUE #11493: Add banner to user and developer facing AMO pages
Until we have a fix for the certificate issue, let's put up a banner on all AMO pages that face users and developers that says: 

.You may be experiencing issues with extensions and themes for Firefox. Please see https://discourse.mozilla.org/t/certificate-issue-causing-add-ons-to-be-disabled-or-fail-to-install/39047 for more information."

### PR DESCRIPTION
**Pull request by eviljeff**
fixes #11493

### CODE REVIEW

File: docs/topics/api/overview.rst [hunk 1]
```diff
@@ -246,6 +249,25 @@ specified.
...
 .. _api-versions-list:
 
+
+-----------
+Site Status
+-----------
+
+.. _`api-site-status`:
+
+This special endpoint just returns if the site is in read only mode, and if there is a site notice currently in effect.
```
Comment (willdurand) [inline docs/topics/api/overview.rst]: Not sure I fully understand this sentence.

File: docs/topics/api/overview.rst [hunk 2]
```diff
@@ -246,6 +249,25 @@ specified.
...
+See :ref:`maintainance mode <api-overview-maintainance>` for mode details of when the site is read only and how requests are affected.
```
Comment (willdurand) [inline docs/topics/api/overview.rst]:
```suggestion
See :ref:`maintainance mode <api-overview-maintainance>` for more details of when the site is read only and how requests are affected.
```

File: docs/topics/api/overview.rst [hunk 3]
```diff
@@ -111,6 +113,7 @@ The response when returning ``HTTP 503 Service Unavailable`` in case of read-onl
     }
 
 In case we are not in read-only mode everything should be back working as normal.
+To check if the site is in read-only mode before submitting a response the :ref:`site status api<api-site-status>` can be checked.
```
Comment (willdurand) [inline docs/topics/api/overview.rst]:
```suggestion
To check if the site is in read-only mode before submitting a response, the :ref:`site status api<api-site-status>` can be called.
```

File: docs/topics/api/overview.rst [hunk 4]
```diff
@@ -246,6 +249,25 @@ specified.
...
+
+This special endpoint just returns if the site is in read only mode, and if there is a site notice currently in effect.
+See :ref:`maintainance mode <api-overview-maintainance>` for mode details of when the site is read only and how requests are affected.
+
+
+.. http:get:: /api/v4/site/
+
+    .. _site-status-object:
+
+    :>json boolean read_only: Is the site in read-only mode.
```
Comment (willdurand) [inline docs/topics/api/overview.rst]:
```suggestion
    :>json boolean read_only: Whether the site in read-only mode.
```

File: docs/topics/api/overview.rst [hunk 5]
```diff
@@ -246,6 +249,25 @@ specified.
...
+    :>json string notice: A site-wide notice about any current known difficulties or restrictions.  If this API is being consumed by a tool/frontend it should be displayed to the user.
```
Comment (willdurand) [inline docs/topics/api/overview.rst]: Will the message be localized?

File: src/olympia/accounts/tests/test_serializers.py [hunk 6]
```diff
@@ -250,6 +251,21 @@ def test_validate_homepage(self):
...
             assert serializer.validate_homepage(allowed_url) == allowed_url
 
+    def test_site_status(self):
+        data = super(TestUserProfileSerializer, self).test_basic()
+        assert data['site_status'] == {
+            'read_only': False,
+            'notice': None,
+        }
+
+        set_config('site_notice', 'THIS is NOT Á TEST!')
```
Comment (willdurand) [inline src/olympia/accounts/tests/test_serializers.py]: We could maybe have one `assert` here to cover the case where the site is not in read only mode but there is a message to display to the users.

File: src/olympia/amo/views.py [hunk 7]
```diff
@@ -131,3 +134,15 @@ def _frontend_view(*args, **kwargs):
...
+class SiteStatusView(APIView):
+    authentication_classes = []
+    permission_classes = []
+
+    @classmethod
+    def as_view(cls, **initkwargs):
+        return non_atomic_requests(super().as_view(**initkwargs))
+
+    def get(self, request, format=None):
+        return Response(SiteStatusSerializer(object()).data)
```
Comment (willdurand) [inline src/olympia/amo/views.py]: Why do we need `object()` here?

File: src/olympia/accounts/serializers.py [hunk 8]
```diff
@@ -79,12 +80,13 @@ class UserProfileSerializer(PublicUserProfileSerializer):
     reviewer_name = serializers.CharField(
         min_length=2, max_length=50, allow_blank=True,
         validators=[OneOrMorePrintableCharacterAPIValidator()])
+    site_status = SiteStatusSerializer(source='*')
```
Comment (willdurand) [inline src/olympia/accounts/serializers.py]: Why do we need a `source='*'` here?

File: docs/topics/api/overview.rst (hunk 5 above)
Comment (eviljeff) [reply docs/topics/api/overview.rst]: No.

File: docs/topics/api/overview.rst [hunk 9]
```diff
@@ -246,6 +249,25 @@ specified.
...
+This special endpoint just returns if the site is in read only mode, and if there is a site notice currently in effect.
+See :ref:`maintainance mode <api-overview-maintainance>` for more details of when the site is read only and how requests are affected.
+
+
+.. http:get:: /api/v4/site/
+
+    .. _site-status-object:
+
+    :>json boolean read_only: Is the site in read-only mode.
+    :>json string notice: A site-wide notice about any current known difficulties or restrictions.  If this API is being consumed by a tool/frontend it should be displayed to the user.
```
Comment (willdurand) [inline docs/topics/api/overview.rst]: This should probably return `null` when it is not set, no?

File: src/olympia/amo/views.py (hunk 7 above)
Comment (eviljeff) [reply src/olympia/amo/views.py]: It's just how serializers are written - SiteStatusSerializer isn't really serializing input or retrieved data in the strictest sense, and serializers are supposed to be serializing _something_. You either have to pass an instance of an object (an empty object is the most trivial example), or pass data, but then you have to call validate too (and extra few lines of code)

File: src/olympia/accounts/serializers.py (hunk 8 above)
Comment (eviljeff) [reply src/olympia/accounts/serializers.py]: similar to the `object()` above - when a serializer is used as a field inside a serializer it's passed some data from the parent object/data, and by default it's the name of the field.  As `UserProfile` doesn't contain a `site_status` property it fails so you have to give it a source of another field.  `source='*'` is just passing the entire UserProfile.  Could be any property.

File: src/olympia/amo/views.py (hunk 7 above)
Comment (willdurand) [reply src/olympia/amo/views.py]: ah, ok I see now, thanks

File: src/olympia/accounts/serializers.py (hunk 8 above)
Comment (willdurand) [reply src/olympia/accounts/serializers.py]: I see, thanks. Could you maybe add a comment above this line to explain that?

File: docs/topics/api/overview.rst (hunk 1 above)
Comment (EnTeQuAk) [inline docs/topics/api/overview.rst]: This kinda sounds like it'll be blocking if the site is not in read-only mode which isn't quite what you wanted to say I suppose. I wonder if simply removing the "just" makes it much clearer already.

File: src/olympia/amo/tests/test_views.py [hunk 10]
```diff
@@ -443,3 +445,27 @@ def test_version_json(self):
...
+            'notice': None,
+        }
+
+        set_config('site_notice', 'THIS is NOT Á TEST!')
+        with override_settings(READ_ONLY=True):
+            response = self.client.get(self.url)
+        assert response.data == {
+            'read_only': True,
+            'notice': 'THIS is NOT Á TEST!',
+        }
```
Comment (EnTeQuAk) [inline src/olympia/amo/tests/test_views.py]: Let's add two more test-cases that check for `READ_ONLY=False` but `site_notice` set and `READ_ONLY=True` but `site_notice` unset.

File: src/olympia/accounts/tests/test_serializers.py (hunk 6 above)
Comment (eviljeff) [reply src/olympia/accounts/tests/test_serializers.py]: https://github.com/mozilla/addons-server/pull/12356/commits/4acd93a499ade786efc4967b45807f2550d4a248

File: src/olympia/accounts/serializers.py (hunk 8 above)
Comment (eviljeff) [reply src/olympia/accounts/serializers.py]: https://github.com/mozilla/addons-server/pull/12356/commits/4acd93a499ade786efc4967b45807f2550d4a248

File: docs/topics/api/overview.rst (hunk 9 above)
Comment (eviljeff) [reply docs/topics/api/overview.rst]: https://github.com/mozilla/addons-server/pull/12356/commits/4acd93a499ade786efc4967b45807f2550d4a248

File: src/olympia/amo/tests/test_views.py (hunk 10 above)
Comment (eviljeff) [reply src/olympia/amo/tests/test_views.py]: https://github.com/mozilla/addons-server/pull/12356/commits/4acd93a499ade786efc4967b45807f2550d4a248

File: docs/topics/api/overview.rst (hunk 1 above)
Comment (eviljeff) [reply docs/topics/api/overview.rst]: https://github.com/mozilla/addons-server/pull/12356/commits/4acd93a499ade786efc4967b45807f2550d4a248

**Pull request merged**